test/python/compiler/test_scheduler.py

```diff
@@ -0,0 +1,91 @@
+
+"""Scheduler Test."""
+
+import unittest
+
+from qiskit.circuit import QuantumRegister, ClassicalRegister, QuantumCircuit
+from qiskit.exceptions import QiskitError
+from qiskit.pulse import InstructionScheduleMap, Schedule
+from qiskit.test import QiskitTestCase
+from qiskit.test.mock import FakeOpenPulse3Q
+from qiskit.compiler.scheduler import schedule
+
+
+class TestCircuitScheduler(QiskitTestCase):
+    """Tests for scheduling."""
+
+    def setUp(self):
+        super().setUp()
+        qr = QuantumRegister(2, name="q")
+        cr = ClassicalRegister(2, name="c")
+        self.circ = QuantumCircuit(qr, cr, name="circ")
+        self.circ.cx(qr[0], qr[1])
+        self.circ.measure(qr, cr)
+
+        qr2 = QuantumRegister(2, name="q")
+        cr2 = ClassicalRegister(2, name="c")
+        self.circ2 = QuantumCircuit(qr2, cr2, name="circ2")
+        self.circ2.cx(qr2[0], qr2[1])
+        self.circ2.measure(qr2, cr2)
+
+        self.backend = FakeOpenPulse3Q()
+        self.backend_config = self.backend.configuration()
+        self.num_qubits = self.backend_config.n_qubits
+
+    def test_instruction_map_and_backend_not_supplied(self):
+        """Test instruction map and backend not supplied."""
+        with self.assertRaisesRegex(
+            QiskitError,
+            r"Must supply either a backend or InstructionScheduleMap for scheduling passes.",
+        ):
+            schedule(self.circ)
+
+    def test_instruction_map_and_backend_defaults_unavailable(self):
+        """Test backend defaults unavailable when backend is provided, but instruction map is not."""
+        self.backend._defaults = None
+        with self.assertRaisesRegex(
+            QiskitError, r"The backend defaults are unavailable. The backend may not support pulse."
+        ):
+            schedule(self.circ, self.backend)
+
+    def test_measurement_map_and_backend_not_supplied(self):
+        """Test measurement map and backend not supplied."""
+        with self.assertRaisesRegex(
+            QiskitError,
+            r"Must supply either a backend or a meas_map for scheduling passes.",
+        ):
+            schedule(self.circ, inst_map=InstructionScheduleMap())
+
+    def test_schedules_single_circuit(self):
+        """Test scheduling of a single circuit."""
+        circuit_schedule = schedule(self.circ, self.backend)
+
+        self.assertIsInstance(circuit_schedule, Schedule)
+        self.assertEqual(circuit_schedule.name, "circ")
+
+    def test_schedules_multiple_circuits(self):
+        """Test scheduling of multiple circuits."""
+        circuits = [self.circ, self.circ2]
+        circuit_schedules = schedule(circuits, self.backend)
+        self.assertEqual(len(circuit_schedules), len(circuits))
+
+        circuit_one_schedule = circuit_schedules[0]
+        circuit_two_schedule = circuit_schedules[1]
+
+        self.assertEqual(circuit_one_schedule.name, "circ")
+        self.assertEqual(circuit_two_schedule.name, "circ2")
```

Comment:
I think that it'd be good to have some more validation on the output here to ensure that the scheduler config is respected across the parallel dispatch. For example adding some assertions checking the output schedules are both alap or asap scheduled would be good.

Reply:
From what I see looking at the code and experimenting, `qiskit.pulse.Schedule` objects do not have an attribute which identifies the scheduling method used. The object only consists of a name, a set of pulse instructions, and a metadata dictionary (which doesn't contain the scheduling method either).

This leads me to believe that we aren't going to be able to validate the output schedules much more unless we assert that a specific set of instructions are returned within a schedule. I can imagine that being brittle as the scheduling code changes, however. What do you think?

Reply:
I was thinking more that you schedule the two circuits in isolation with the same options and then just compare that the schedules are the same since the scheduling should be deterministic. For example something like:
```python
self.assertEqual(circuit_one_schedule, schedule(self.circ, self.backend, method='alap'))
self.assertEqual(circuit_two_schedule, schedule(self.circ2, self.backend, method='alap'))
```

Reply:
Ah, I see. That's a good idea. Implemented in [18ad153](https://github.com/Qiskit/qiskit-terra/pull/8110/commits/18ad15354f48f261ff1373f87bfdfa1353afdf9d).